File: xxf_http/src/main/java/com/xxf/arch/http/interceptor/HttpExceptionFeedInterceptor.java
package com.xxf.arch.http.interceptor;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;

/**
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @Description 专注日常http异常汇报, 由于header太多, 日志系统显示不全, 所以汇报精简日志
 */
public abstract class HttpExceptionFeedInterceptor implements Interceptor {
    public static final Charset UTF8 = Charset.forName("UTF-8");

    @NotNull
    @Override
    public final Response intercept(@NotNull Interceptor.Chain chain) throws IOException {
        long startNs = System.nanoTime();
        Request request = chain.request();
        Response response;
        try {
            response = chain.proceed(request);
        } catch (Exception e) {
            onFeedHttpException(String.valueOf(request.url()), request.method(), e);
            throw e;
        }
        long tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
        //汇报异常
        if (response != null
                && !response.isSuccessful()) {
            onFeedResponseException(
                    String.valueOf(request.url()),
                    request.method(),
                    getLoggerRequestBody(request.body()),
                    response.code(),
                    response.message(),
                    tookMs,
                    getLoggerResponseBody(response.body()));
        }
        return response;
    }

    /**
     * 异常汇报
     *
     * @param url
     * @param method
     * @param e
     */
    protected abstract void onFeedHttpException(String url, String method, Throwable e);

    /**
     * 响应异常汇报
     *
     * @param url
     * @param method
     * @param reqBody
     * @param code
     * @param message
     * @param tookMs  消耗时间
     * @param resBody
     */
    protected abstract void onFeedResponseException(String url, String method, String reqBody, int code, String message, long tookMs, String resBody);

    /**
     * 获取 响应体
     *
     * @param responseBody
     * @return
     */
    public static String getLoggerResponseBody(ResponseBody responseBody) {
        if (responseBody != null) {
            try {
                long contentLength = responseBody.contentLength();

                BufferedSource source = responseBody.source();
                source.request(Long.MAX_VALUE); // Buffer the entire body.
                Buffer buffer = source.buffer();

                Charset charset = UTF8;
                MediaType contentType = responseBody.contentType();
                if (contentType != null) {
                    charset = contentType.charset(UTF8);
                }

                if (!HttpLoggingInterceptor.isPlaintext(buffer)) {
                    return "binary " + buffer.size() + "-byte body omitted";
                }

                if (contentLength != 0) {
                    return buffer.clone().readString(charset);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return null;
    }

    /**
     * 获取请求体
     *
     * @param requestBody
     * @return
     */
    public static String getLoggerRequestBody(RequestBody requestBody) {
        if (requestBody != null) {
            try {
                Buffer buffer = new Buffer();
                requestBody.writeTo(buffer);

                Charset charset = UTF8;
                MediaType contentType = requestBody.contentType();
                if (contentType != null) {
                    charset = contentType.charset(UTF8);
                }

                if (HttpLoggingInterceptor.isPlaintext(buffer)) {
                    return buffer.readString(charset);
                } else {
                    return "binary " + requestBody.contentLength() + "-byte body omitted";
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return null;
    }
}

File: lib_http/src/main/java/com/xxf/arch/http/interceptor/HttpExceptionFeedInterceptor.kt
package com.xxf.arch.http.interceptor

import android.util.Log
import com.xxf.arch.http.interceptor.internal.isProbablyUtf8
import kotlin.Throws
import okhttp3.*
import okhttp3.internal.http.promisesBody
import okio.Buffer
import okio.GzipSource
import java.io.IOException
import java.lang.StringBuilder
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit

/**
 * @Author: XGod  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq  xuanyouwu@163.com  17611639080  https://github.com/NBXXF     https://blog.csdn.net/axuanqq
 * @Description 专注日常http异常汇报, 由于header太多, 日志系统显示不全, 所以汇报精简日志
 */
open class HttpExceptionFeedInterceptor : Interceptor {
    @Throws(IOException::class)
    final override fun intercept(chain: Interceptor.Chain): Response {
        val request: Request = chain.request()
        val startNs = System.nanoTime()
        val response: Response = try {
            chain.proceed(request)
        } catch (e: Throwable) {
            val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
            onFeedHttpException(request, response = null, throwable = e, tookMs = tookMs)
            throw e
        }
        val tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs)
        onFeedHttpException(request, response = response, throwable = null, tookMs)
        return response
    }

    /**
     * 发现异常
     */
    open fun onFeedHttpException(
        request: Request,
        response: Response?,
        throwable: Throwable?,
        tookMs: Long
    ) {
        if (throwable != null || (response != null && !response.isSuccessful)) {
            val sb = StringBuilder("http fail")
            sb.append("\n")
            sb.append("requestUrl:[${request.method}] - ${request.url} - (${tookMs}ms)")
            sb.append("\n")
            sb.append("requestHeaders:\n${request.headers}")
            if (request.body != null) {
                sb.append("requestBody:\n${logRequestBody(request)}")
            } else {
                sb.append("requestBody:")
            }
            sb.append("\n")
            sb.append("responseHeaders:\n${response?.headers}")
            sb.append("responseBody:\n${response?.let { logResponseBody(it) }}")
            sb.append("\n")
            sb.append("throwable:${Log.getStackTraceString(throwable)}")
            onFeedHttpException(sb.toString())
        }
    }

    open fun onFeedHttpException(string: String) {

    }

    private fun logRequestBody(request: Request): String {
        val requestBody = request.body ?: return ""
        when {
            bodyHasUnknownEncoding(request.headers) -> {
                return "(encoded body omitted)"
            }
            requestBody.isDuplex() -> {
                return "(duplex request body omitted)"
            }
            requestBody.isOneShot() -> {
                return "(one-shot body omitted)"
            }
            else -> {
                val buffer = Buffer()
                requestBody.writeTo(buffer)

                val contentType = requestBody.contentType()
                val charset: Charset =
                    contentType?.charset(StandardCharsets.UTF_8) ?: StandardCharsets.UTF_8

                return if (buffer.isProbablyUtf8()) {
                    buffer.readString(charset)
                } else {
                    "body omitted"
                }
            }
        }
    }

    private fun logResponseBody(response: Response): String {
        if (!response.promisesBody()) {
            return "<-- END HTTP"
        } else if (bodyHasUnknownEncoding(response.headers)) {
            return "<-- END HTTP (encoded body omitted)"
        } else {
            val responseBody = response.body ?: return ""
            val source = responseBody.source()
            source.request(Long.MAX_VALUE) // Buffer the entire body.
            var buffer = source.buffer

            val headers = response.headers
            var gzippedLength: Long? = null
            if ("gzip".equals(headers["Content-Encoding"], ignoreCase = true)) {
                gzippedLength = buffer.size
                GzipSource(buffer.clone()).use { gzippedResponseBody ->
                    buffer = Buffer()
                    buffer.writeAll(gzippedResponseBody)
                }
            }

            val contentType = responseBody.contentType()
            val charset: Charset =
                contentType?.charset(StandardCharsets.UTF_8) ?: StandardCharsets.UTF_8

            if (!buffer.isProbablyUtf8()) {
                return "(binary ${buffer.size}-byte body omitted)"
            }

            val contentLength = responseBody.contentLength()
            if (contentLength != 0L) {
                return buffer.clone().readString(charset)
            }
            return if (gzippedLength != null) {
                "$gzippedLength-gzipped-byte body)"
            } else {
                "${buffer.size}-byte body)"
            }
        }
    }

    private fun bodyHasUnknownEncoding(headers: Headers): Boolean {
        val contentEncoding = headers["Content-Encoding"] ?: return false
        return !contentEncoding.equals("identity", ignoreCase = true) &&
                !contentEncoding.equals("gzip", ignoreCase = true)
    }
}